File: Algorithms/Graph_Algorithms/DFS/dfs.js
function dfs(at) {
  // if we have been visited the node then return
  if(visited[at]) return;
  // mark the node as visited
  visited[at] = true;
  
  result.push(at); // or you can do something with the node

  // find the nodes in depth
  let nodes = adjList.get(at);
  for (let neigh of nodes) {
    dfs(neigh);
  }
}

function addNode(node) {
  // adding the node, at first point to nothing
  adjList.set(node, []);
}

function addEdge(origin, destination) {
  // finding the node origin and set the edge to destination
  adjList.get(origin).push(destination);
  // finding the node destination and set the edge to origin
  adjList.get(destination).push(origin);
}

// T E S T
const nodes = [0,1,2,3,4,5,6,7,8,9,10];
const edges = [
  [0, 4],
  [0, 3],
  [3, 2],
  [3, 7],
  [3, 8],
  [5, 4],
  [5, 1],
  [5, 10],
  [5, 6],
  [10, 1],
];

// define a structure to support the graph
let adjList = new Map();

// biulding the nodes of the graph
nodes.forEach(addNode);

// building the edges
edges.forEach(edge=>addEdge(...edge));

// building an array for the flag of the visited nodes
let visited = new Array(nodes.length).fill(false);

// traverse starting from node 0
let start_node = 0;
let result = [];
dfs(start_node);
console.log(`the result of traverse the graph from ${start_node} is`, result);

File: Algorithms/Graph_Algorithms/DFS/dfs.c
// Depth First Search
// C Programming

#include <stdio.h>
#include <stdlib.h>

// Node -> A structure with a vertex and a link to the next Node.
struct Node {
  int vertex;
  struct Node* nextNode;
};

// Graph       -> A structure to represent a Graph.
// numVertices -> Number of Vertices.
// visited     -> Visited Nodes List 
// adjLists    -> To store an array of Linked lists
struct Graph {
  int numVertices;
  int* visited;
  struct Node** adjLists;
};

// Function name : createNode
// Input         : int v
// Output        : struct* Node
// Logic         : For creating a new Node by allocating memory for it.
struct Node* createNode(int v) {
  struct Node* newNode = malloc(sizeof(struct Node));
  newNode->vertex = v;
  newNode->nextNode = NULL;
  return newNode;
}

// Function name : DFS
// Input         : Graph graph, int vertex
// Output        : void
// Logic         : For implementing the Depth First Search Algorithm.
void DFS(struct Graph* graph, int vertex) {
  struct Node* adjList = graph->adjLists[vertex];
  struct Node* temp = adjList;

  graph->visited[vertex] = 1;
  printf("Visited %d \n", vertex);

  while (temp != NULL) {
    int connectedVertex = temp->vertex;

    if (graph->visited[connectedVertex] == 0) {
      DFS(graph, connectedVertex);
    }
    temp = temp->nextNode;
  }
}



// Function name : createGraph
// Input         : int vertices
// Output        : struct Graph
// Logic         : For creating Graphs.
struct Graph* createGraph(int vertices) {
  struct Graph* graph = malloc(sizeof(struct Graph));
  graph->numVertices = vertices;

  graph->adjLists = malloc(vertices * sizeof(struct Node*));

  graph->visited = malloc(vertices * sizeof(int));

  int i;
  for (i = 0; i < vertices; i++) {
    graph->adjLists[i] = NULL;
    graph->visited[i] = 0;
  }
  return graph;
}

// Function name : addEdge
// Input         : struct Graph, int src, int dest
// Output        : void
// Logic         : For adding an edge to the graph
void addEdge(struct Graph* graph, int src, int dest) {
  // Add edge from src to dest
  struct Node* newNode = createNode(dest);
  newNode->nextNode = graph->adjLists[src];
  graph->adjLists[src] = newNode;

  // Add edge from dest to src
  newNode = createNode(src);
  newNode->nextNode = graph->adjLists[dest];
  graph->adjLists[dest] = newNode;
}

// Function name : printGraph
// Input         : struct Graph
// Output        : void
// Logic         : For printing the Graph.
void printGraph(struct Graph* graph) {
  int v;
  for (v = 0; v < graph->numVertices; v++) {
    struct Node* temp = graph->adjLists[v];
    printf("\n Adjacency list of vertex %d\n ", v);
    while (temp) {
      printf("%d -> ", temp->vertex);
      temp = temp->nextNode;
    }
    printf("\n");
  }
}

int main() {
    int nVertices, nEdges, srcNode, destNode, vertex;

    printf("\nNumber of vertices : ");
    scanf("%d", &nVertices);

    struct Graph* graph = createGraph(nVertices);
    
    printf("\nNumber of edges : ");
    scanf("%d", &nEdges);

    for(int i=0; i<nEdges; i++){
        printf("Edge : ");
        scanf("%d %d", &srcNode, &destNode);
        addEdge(graph, srcNode, destNode);
    }

    printGraph(graph);

    printf("\nVertex to search for : ");
    scanf("%d", &vertex);

    DFS(graph, vertex);

    return 0;
}